### DISCUSSION TITLE
Simple button is not working

### DISCUSSION THREAD

**Discussion by greenreader9** [micropython/micropython#9328] [Using MicroPython] [answered]
Hello!

I am fairly new to microcontrollers, but have successfully programed a short script (CircuitPython) in the past, and I have absolutely no clue what is going on here (With MicroPython).

The board is a Raspberry Pi Pico W running version `20220914-unstable-v1.19.1-409-g0e8c2204d` installed with Thonny (See screenshot below)

![image](https://user-images.githubusercontent.com/75644395/190520869-fa82dc66-a2b4-4804-8d9a-fc65f2810f6e.png)

My code is:

```py
from machine import Pin

btn = Pin(0, Pin.IN)

while True:
    if btn() == 1:
        print("pushed")
    else:
        print("Not pushed")
```

[Looking at this image](https://th.bing.com/th/id/R.7b55c0d845f4d4a3cab0c63586571cbe?rik=9CmOGnFr2i6E7w&ehk=lFa6sw%2bavwi65sfpHf0n1YmUuTHrxbFs5vTxi1YdNGQ%3d), I am using 1 ( `GPO 0`) and 36 (`3V3(OUT)`).

Since the button does not seem to work right at all, I just removed it and tried the below to see how the output changes when I changed the input.

So here are the four things I have done, and what happens when I do them:

1) Nothing (So just run the program with nothing attached): `Not Pushed` is printed repeatedly (So this works)
2) Connect `GPO 0` to `3V3(OUT)` with a jumper cable: `Pushed` is printed repeatedly (So this works)
3) Start with nothing connected, then connect `GPO 0` to `3V3(OUT)` with a jumper cable: `Not Pushed` is printed repeatedly (So this does not work right, since it should switch to `Pushed` when I add in the jumper cable)
4) Start with `GPO 0` and `3V3(OUT)` connected with a jumper cable, then disconnect them: `Pushed` is printed repeatedly  (So this does not work right, since it should switch to `Not Pushed` when I remove the jumper cable)

So for some reason it always just prints whatever it's initial state was, why is this?

Apologies if the answer is really really simple, but nothing I find on the internet seems to work at all.

Thanks in advance!

**Comment by davefes** [comment]
Try changing:
```py
if btn() == 1:
```
to 
```py
if btn.value() == 1:
```

**Reply by greenreader9** [reply]
No, that did not work. Still the same behavior unfortunately.

**Comment by davefes** [comment]
OK, I am working through a similar issue where GPIO state is unchanged on doing a CTRL-C in rshell.

Something else to try before your while loop:
```py
btn.off()
```
then at least you know what state you are starting off in.

**Reply by greenreader9** [reply]
Assuming this is what you meant:
```py
from machine import Pin

btn = Pin(0, Pin.IN)
btn.off()

while True:
    if btn.value() == 1:
        print("pushed")
    else:
        print("Not pushed")
```

It still does not work, returning the same results as before

**Comment by davefes** [comment]
After 
```py
btn.off()
```
put
```py
print (' btn state = ', btn.value())
```
Sorry, I don't use Thonny so maybe I am missing something there.

**Comment by robert-hh** [comment]
Do you have an external pull resistor to define the button open state? You do not use the `pull=` in your code. If the button connects the input to Vcc, you need a pull-down default, if it connects to GND, you need pull-up. You can at least try with the internal pull options:
`btn = Pin(0, Pin.IN, pul=Pin.PULL_DOWN)` for a pull-down or `btn = Pin(0, Pin.IN, pull=PULL_UP)` for pull-up.

**Comment by steveh127** [comment]
I can confirm Roberts code works as expected or almost, there are a couple of obvious and minor typos in Roberts code. `btn=Pin(0, Pin.IN, pull=Pin.PULL_DOWN )` and you don't need the `pull=` so `btn=Pin(0,Pin.IN,Pin.PULL_UP) ` works too. Tested on a Pico W.

**Reply by greenreader9** [reply]
Now using this:

```py
from machine import Pin

btn=Pin(0, Pin.IN, pull=Pin.PULL_DOWN )
btn.off()

while True:
    print (' btn state = ', btn.value())
```

Still has the same behavior as before...

**Comment by greenreader9** [comment, marked as answer]
I figured it out.

For future reference, this worked for me:

```py
from machine import Pin

btn = machine.Pin(0, machine.Pin.IN, machine.Pin.PULL_DOWN)

last_state = False
current_state = False

while True:
    current_state = btn.value()
    if last_state == False and current_state == True:
        print ('pushed')
    last_state = current_state
```

Thanks to everyone that replied!

**Reply by puppet13th** [reply]
there is 2 methods you can use for simple button:
```
machine.Pin.PULL_DOWN ==> set default pin state to low
machine.Pin.PULL_UP ==> set default pin state to high
```
PULL_DOWN scenario
wiring : [Pin.IN] -- [button/swtich] -- [3.3v]
not pressed value = 0
pressed value = 1
```py
while True:
    if btn.value() == 1:
        print('button pressed')
```
PULL_UP scenario
wiring : [Pin.IN] -- [button/swtich] -- [GND]
not pressed value = 1
pressed value = 0
```py
while True:
    if btn.value() == 0:
        print('button pressed')

**Comment by jimmo** [comment]
Just FYI @davefes @greenreader9 -- `pin()` and `pin.value()` are equivalent. The former is marginally more efficient in terms of how the VM executes it, but I think the latter is clearer.
(In Python terminology, `machine.Pin` objects implement the `__call__` operator).
